### ISSUE TITLE
MethodHandle interface to ObjectInstantiator

### ISSUE THREAD

**Issue by stepancheg** [easymock/objenesis#42]
`MethodHandle` is a JDK interface to generate fast code on the fly.

Suppose, you want to have to impement operation like this:

* instantiate an object
* populate some fields of that object

With help of objenesis it could be implemented like this:

```
Object my = objectInstantiator.newInstance();
populateMethodHandle.invokeExact(my);
```

It is not very efficient, because `populateMethodHandle.invokeExact` performs signature check on each call (unless it is stored in static final field, which is not always possible), and because `objectInstaniator.newInstance` also performs reflective calls.

However, if `ObjectInstantiator` could expose itself as `MethodHandle`, these two operations could be combned into one.

```
MethodHandle newInstanceAndPopulate = MethodHandles.filterReturnValue(
    objectInstantiator.newInstanceMethodHandle(),
    populateMethodHandle);

...

Object my = newInstanceAndPopulate.invokeExact();
```

As far as I understand, it is possible to expose `MethodHandle` from some instantiators, for example, from `MagicInstantiator` and probably from `SunReflectionFactoryInstantiator` (with `MethodHandles.Lookup.unreflectConstructor(..)`).

`MethodHandle newInstanceMethodHandle()` method could be added to `ObjectInstantiator` interface (probably as returning `Object` for compatibility with older JDK), or method could be added to additional interface like `MethodHandleAwareObjectInstantiator extends ObjectInstantiator`.

**Comment by henri-tremblay** [comment]
I have some questions about that:

1. Why are you using Objenesis instead of just calling a constructor taking the parameters you want to set?
2. You are using the method handle with generated code? You are not calling and actual method on the class? (I'm not a MethodHandle expert so this sentence might makes no sense)
3. Why is the filter a better way?

**Comment by stepancheg** [comment]
> Why are you using Objenesis instead of just calling a constructor taking the parameters you want to set?

Because constructor may be private, and because constructor may have side-effects which are not acceptable in certain use cases (see example below).

> You are using the method handle with generated code?
> Why is the filter a better way?

Classes are hand-written, but used with `MethodHandle`. My actual use case is this (simplified): user provides a "value" type, like `class Apple { float weight; String color }`. Framework creates an `ArrayList`-like structure which scatters fields of the class to arrays (to reduce memory consumption and GC overhead):

```
// `A` is a "value" type like `Apple`
class UnpackedArrayList<A> {
    // Class (probably generated) containing `float[]` and `String[]` fields for `Apple`
    Object arrays;
    int size;
    ObjectInstantiator<A> instantiator;
    // This MH is a combination of several other calls
    // which copy elements of arrays to fields of A
    MethodHandle populateElementFromArrays;

    A get(int index) {
        // `Apple` class may have no constructor
        A result = instantiator.newInstance();

        // Populate fields of `result` object from values in `arrays`.
        // Method handles are super-convenient here, because 
        // unlike reflection, several field operations are glued into one,
        // and runtime typechecking happens only once, during `invokeExact`
        populateElementFromArrays.invokeExact(arrays, index, result);

        // If `instantiator` could provide a `MethodHandle` interface,
        // instantiator+populate could be glued into single `MethodHandle`
        // (with `MethodHandles.filter`), so
        // there wouldn't be a runtime type check of `A` class
        // between first and second operation.

        return result;
    }
}
```

> You are not calling and actual method on the class? (I'm not a MethodHandle expert so this sentence might makes no sense)

I hope, I've explained it above. The idea is that code is generic (reflective), and multiple `MethodHandles` are combined into one to reduce runtime typecheck and boxing overhead.

**Comment by henri-tremblay** [comment]
I will have a look and possibly benchmark it. However, probably not right away. I`m interested in a pull request if you can. The MethodHandleAware flavor.

**Comment by stepancheg** [comment]
I've written a [simple benchmark](https://gist.github.com/stepancheg/437df8c551f5fb75ea162d10797d9fd2) just to demonstrate how MH combination can make reflection faster.

Results on my notebook are:

```
Benchmark                          Mode  Cnt   Score   Error  Units
MhBenchmark.combined               avgt   12  10,074 ± 0,492  ns/op
MhBenchmark.constructorThenMethod  avgt   12  20,029 ± 0,518  ns/op
MhBenchmark.constructorThenMh      avgt   12  17,610 ± 0,392  ns/op
MhBenchmark.manual                 avgt   12   5,588 ± 0,218  ns/op
```

Note that it variables are intentially declared `Object` and not `My` to emulate framework operation, and method handles are not assigned to static final variables (it would be much faster, but it is not always possible).

**Comment by stepancheg** [comment]
Apparently, `SunReflectionFactoryInstantiator` cannot be converted to `MethodHandle` (I just tried, I doesn't work), so the only remaining option is `MagicInstantiator`.

I could be converted to `MethodHandle`, but I think it does not worth all the complications with objenesis compatibility with Java 1.5 (MethodHandles appeared in Java 1.7).

So I think it would be simpler for `MagicInstantiator` to simply expose its internals. I'll open another issue for that.
